### PR TITLE
H3 interop server and assorted fixes

### PR DESCRIPTION
**Pull request by stammw**
I've been working on this `H3` interop server to deploy on h3.stammw.eu.

I'm not sure we'll want to merge this into quinn as it may be out of scope?

The reasons why I think we need to separate h3 interop from the example code are:
- Interop already needs / will need a different set of features. For example transfer tests are based on urls like `https://example.com/100M`
- The example code will be kept simple
- It may be interesting to implement remote logging for people running their interop client against this endpoint. They could filter by IP or CID with a simple web app served via h1/h2.

~For the moment I'm stuck on a bug where requesting more than 1,6M  results in the client missing the tail of a `Data` frame. Receiving the start of a fresh Data frame instead.~

This PR also contains fixes for critical bugs in:
- data writing management
- parsing incomplete frame headers

### CODE REVIEW

**Comment by Ralith** [comment]
> I'm not sure we'll want to merge this into quinn as it may be out of scope?

I think it's appropriate for this to live in the repo so long as people are performing interop tests that benefit from targeted support, which may be indefinitely.

I strongly agree with making this separate from the canonical example server, and I look forward to simplifying the example server (perhaps turning it into an echo server) once this is ready and we can switch over to it for formal testing.

**Comment by djc** [comment]
Fully agree with @Ralith's points.

**Comment by stammw** [comment]
Ready for a review.

I'll get this up and running on my box. Then I think we could publish the endpoint on slack ?

**Comment by stammw** [comment]
Ho, rerty and ZeroRTT aren't actually working...

File: quinn-h3/src/frame.rs [hunk 1]
```diff
@@ -158,12 +158,14 @@ impl Future for WriteFrame {
                 WriteFrameState::Payload(ref mut send, ref mut p) => {
                     let wrote = ready!(Pin::new(send).poll_write(cx, p))?;
                     p.advance(wrote);
-                    let send = match mem::replace(&mut self.state, WriteFrameState::Finished) {
-                        WriteFrameState::Payload(s, _) => s,
-                        _ => unreachable!(),
-                    };
-                    self.state = WriteFrameState::Finished;
-                    return Poll::Ready(Ok(send));
+                    if p.is_empty() {
```
Comment (djc) [inline quinn-h3/src/frame.rs]: Both here and in the preceding block, it might help readability to invert the `is_empty()` condition and do an early return if not empty (this can be a follow-up commit).

Comment (djc) [reply quinn-h3/src/frame.rs]: Also, how hard would it be to add a test that covers this?

Comment (stammw) [reply quinn-h3/src/frame.rs]: It is doable I guess, making a trait out of the used methods of `SendStream` in `WriteFrame`, then using `WriteFrame<T: SendStreamTrait>`. So the tests can pass a write-to-buffer implementation. Some parameters for this mocked impl could then be used to define some `SendStream` desired testing behavior. I'm not sure how hard can this be.

There might be simpler solutions?

Testing all edge cases will be a must-have in the future of this crate .

Comment (stammw) [reply quinn-h3/src/frame.rs]: Do you want those tests included with this PR?

Comment (djc) [reply quinn-h3/src/frame.rs]: Nah, but maybe write an issue about it so we don't forget if you don't intend to do it nowish?
